common/persistence/elasticsearch/processor.go

```diff
@@ -163,40 +164,53 @@ func (p *processorImpl) hashFn(key interface{}) uint32 {
 	return hash % p.indexerConcurrency
 }
 
-// Add request to bulk, and record ack channel message in map with provided key
-func (p *processorImpl) Add(request *client.BulkableRequest, visibilityTaskKey string, ackCh chan<- bool) {
-	if cap(ackCh) < 1 {
-		panic("ackCh must be buffered channel (length should be 1 or more)")
-	}
-	sw := p.metricsClient.StartTimer(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorRequestLatency)
-	ackChWithStopwatch := newAckChanWithStopwatch(ackCh, sw)
-	_, isDup, _ := p.mapToAckChan.PutOrDo(visibilityTaskKey, ackChWithStopwatch, func(key interface{}, value interface{}) error {
-		ackChWithStopwatchExisting, ok := value.(*ackChanWithStopwatch)
+// Add request to the bulk and return ack channel which will receive ack signal when request is processed.
+func (p *processorImpl) Add(request *client.BulkableRequest, visibilityTaskKey string) <-chan bool {
+	ackCh := newAckChan()
+	retCh := ackCh.ackChInternal
+	_, isDup, _ := p.mapToAckChan.PutOrDo(visibilityTaskKey, ackCh, func(key interface{}, value interface{}) error {
+		ackChExisting, ok := value.(*ackChan)
 		if !ok {
-			p.logger.Fatal(fmt.Sprintf("mapToAckChan has item of a wrong type %T (%T expected).", value, &ackChanWithStopwatch{}), tag.Value(key))
+			p.logger.Fatal(fmt.Sprintf("mapToAckChan has item of a wrong type %T (%T expected).", value, &ackChan{}), tag.Value(key))
 		}
 
 		p.logger.Warn("Adding duplicate ES request for visibility task key.", tag.Key(visibilityTaskKey), tag.ESDocID(request.ID), tag.Value(request.Doc))
 
 		// Nack existing visibility task.
-		ackChWithStopwatchExisting.addToProcessStopwatch.Stop()
-		ackChWithStopwatchExisting.ackCh <- false
+		ackChExisting.Done(false, p.metricsClient)
 
 		// Replace existing dictionary item with new item.
 		// Note: request won't be added to bulk processor.
-		ackChWithStopwatchExisting.addToProcessStopwatch = ackChWithStopwatch.addToProcessStopwatch
-		ackChWithStopwatchExisting.ackCh = ackChWithStopwatch.ackCh
+		ackChExisting.addedAt = ackCh.addedAt
+		ackChExisting.startedAt = ackCh.startedAt
+		ackChExisting.ackChInternal = ackCh.ackChInternal
 		return nil
 	})
-	if isDup {
-		return
+	if !isDup {
+		p.bulkProcessor.Add(request)
 	}
-	p.bulkProcessor.Add(request)
+	return retCh
 }
 
 // bulkBeforeAction is triggered before bulk processor commit
 func (p *processorImpl) bulkBeforeAction(_ int64, requests []elastic.BulkableRequest) {
-	p.metricsClient.AddCounter(metrics.ElasticSearchVisibility, metrics.ESBulkProcessorRequests, int64(len(requests)))
+	p.metricsClient.AddCounter(metrics.ElasticsearchBulkProcessor, metrics.ElasticsearchBulkProcessorRequests, int64(len(requests)))
+	p.metricsClient.RecordDistribution(metrics.ElasticsearchBulkProcessor, metrics.ElasticsearchBulkProcessorBulkSize, len(requests))
+
+	for _, request := range requests {
+		visibilityTaskKey := p.extractVisibilityTaskKey(request)
+		if visibilityTaskKey == "" {
+			continue
+		}
```

Comment:
which case will have `visibilityTaskKey == ""`

Reply:
When `extractVisibilityTaskKey` wasn't able to extract it. It doesn't return error because it already logged it and reported metrics. So here I just ignore the item.

Reply:
I know everyone feels code should be self-documenting (and it should whenever possible) but personally I think that when an experienced reviewer familiar with the code has a question about what's going on, that's a good sign that a comment might be worthwhile.  Just saying :)